Native compilation runtime error: java.lang.ClassNotFoundException: org.apache.kafka.common.security.scram.internals.ScramSaslClient$ScramSaslClientFactory

**Describe the bug**
While developing a Spring Boot microservice that uses Spring Cloud Stream Kafka Binder, as reported in this issue on Spring Cloud Stream repo:

https://github.com/spring-cloud/spring-cloud-stream/issues/2968

with this extract of application.yml configuration:

```yaml
cloud:
    refresh:
      enabled: false
      client:
        refresh-enabled: false
        config:
          default:
            logger-level: full
    stream:
      bindings:
        kafkaMessageConsumer-in-0:
          destination: devicemanager-topic        
      kafka:
        bindings:
          kafkaMessageConsumer-in-0:
            consumer:
              configuration:
                value.deserializer: org.springframework.kafka.support.serializer.JsonDeserializer
                spring.json.value.default.type: com.minsait.phygital.resource.devicemanager.outadapter.domain.MetricsMessageBean
        binder:
            brokers: kafka-test.iiot.ourbroker.com:9098
            configuration:
              commit.interval.ms: 100
              default:
                value.deserializer: org.springframework.kafka.support.serializer.JsonDeserializer
                spring.json.value.default.type: com.minsait.phygital.resource.devicemanager.outadapter.domain.MetricsMessageBean
              security:
                protocol: SASL_SSL
              sasl:
                mechanism: SCRAM-SHA-512
                jaas:
                  config: org.apache.kafka.common.security.scram.ScramLoginModule required username=test password=test-secret;
....... cutted......
```

I was experiencing a runtime error reported below in the logs only on the native app version.

@sobychacko pointed me that this is due to missing reflection metadata for the apache kafka client that is used by a Spring Cloud Stream Kafka binder. He suggested me to provide a fix in this repo I will do shortly after this issue with a PR

**To Reproduce**

- Create a Spring Cloud Stream app with the above part in your application.yml as cloud stream config using the kafka binder to connect to a kafka broker.
- Compile and run on your JRE to see no issue
- Compile native and run the native executable to experience the reported issue

```bash
mvn clean package -Pnative
```

**Expected behavior**

The native version should run without issues like the ClassNotFoundException one reported below. A local fix on my app is to add this two native hints to run without issues:

```java
      hints.reflection().registerType(
          org.apache.kafka.common.security.scram.internals.ScramSaslClient.ScramSaslClientFactory.class,
          builder -> builder.withMembers(MemberCategory.INVOKE_DECLARED_METHODS,
              MemberCategory.INVOKE_DECLARED_CONSTRUCTORS));
      hints.reflection().registerType(org.apache.kafka.common.security.scram.ScramLoginModule.class,
          builder -> builder.withMembers(MemberCategory.INVOKE_DECLARED_METHODS,
              MemberCategory.INVOKE_DECLARED_CONSTRUCTORS));
```

**Logs**

Launching the native app produces this log:

````
2024-07-01T16:29:13.266+02:00  WARN 10588 --- [| adminclient-1] org.apache.kafka.clients.NetworkClient   : [AdminClient clientId=adminclient-1] Error connecting to node kafka-test.iiot.ourbroker.com:9098 (id: -1 rack: null)

java.io.IOException: Channel could not be created for socket java.nio.channels.SocketChannel[closed]
        at org.apache.kafka.common.network.Selector.buildAndAttachKafkaChannel(Selector.java:348) ~[na:na]
        at org.apache.kafka.common.network.Selector.registerChannel(Selector.java:329) ~[na:na]
        at org.apache.kafka.common.network.Selector.connect(Selector.java:256) ~[na:na]
        at org.apache.kafka.clients.NetworkClient.initiateConnect(NetworkClient.java:1032) ~[na:na]
        at org.apache.kafka.clients.NetworkClient.ready(NetworkClient.java:301) ~[na:na]
        at org.apache.kafka.clients.admin.KafkaAdminClient$AdminClientRunnable.sendEligibleCalls(KafkaAdminClient.java:1109) ~[na:na]
        at org.apache.kafka.clients.admin.KafkaAdminClient$AdminClientRunnable.processRequests(KafkaAdminClient.java:1369) ~[na:na]
        at org.apache.kafka.clients.admin.KafkaAdminClient$AdminClientRunnable.run(KafkaAdminClient.java:1312) ~[na:na]
        at java.base@21.0.2/java.lang.Thread.runWith(Thread.java:1596) ~[device-manager:na]
        at java.base@21.0.2/java.lang.Thread.run(Thread.java:1583) ~[device-manager:na]
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:833) ~[device-manager:na]
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:211) ~[na:na]
Caused by: org.apache.kafka.common.KafkaException: org.apache.kafka.common.errors.SaslAuthenticationException: Failed to configure SaslClientAuthenticator
        at org.apache.kafka.common.network.SaslChannelBuilder.buildChannel(SaslChannelBuilder.java:239) ~[na:na]
        at org.apache.kafka.common.network.Selector.buildAndAttachKafkaChannel(Selector.java:338) ~[na:na]
        ... 11 common frames omitted
Caused by: org.apache.kafka.common.errors.SaslAuthenticationException: Failed to configure SaslClientAuthenticator
Caused by: org.apache.kafka.common.errors.SaslAuthenticationException: Failed to create SaslClient with mechanism SCRAM-SHA-512
Caused by: javax.security.sasl.SaslException: Cannot instantiate service SASL/SCRAM Client Provider: SaslClientFactory.SCRAM-SHA-512 -> org.apache.kafka.common.security.scram.internals.ScramSaslClient$ScramSaslClientFactory

        at java.security.sasl@21.0.2/javax.security.sasl.Sasl.loadFactory(Sasl.java:461) ~[na:na]
        at java.security.sasl@21.0.2/javax.security.sasl.Sasl.createSaslClient(Sasl.java:432) ~[na:na]
        at org.apache.kafka.common.security.authenticator.SaslClientAuthenticator.lambda$createSaslClient$0(SaslClientAuthenticator.java:219) ~[na:na]
        at java.base@21.0.2/java.security.AccessController.executePrivileged(AccessController.java:114) ~[na:na]
        at java.base@21.0.2/java.security.AccessController.doPrivileged(AccessController.java:714) ~[na:na]
        at java.base@21.0.2/javax.security.auth.Subject.doAs(Subject.java:525) ~[device-manager:na]
        at org.apache.kafka.common.security.authenticator.SaslClientAuthenticator.createSaslClient(SaslClientAuthenticator.java:215) ~[na:na]
        at org.apache.kafka.common.security.authenticator.SaslClientAuthenticator.<init>(SaslClientAuthenticator.java:206) ~[na:na]
        at org.apache.kafka.common.network.SaslChannelBuilder.buildClientAuthenticator(SaslChannelBuilder.java:285) ~[na:na]
        at org.apache.kafka.common.network.SaslChannelBuilder.lambda$buildChannel$1(SaslChannelBuilder.java:228) ~[na:na]
        at org.apache.kafka.common.network.KafkaChannel.<init>(KafkaChannel.java:143) ~[na:na]
        at org.apache.kafka.common.network.SaslChannelBuilder.buildChannel(SaslChannelBuilder.java:236) ~[na:na]
        at org.apache.kafka.common.network.Selector.buildAndAttachKafkaChannel(Selector.java:338) ~[na:na]
        at org.apache.kafka.common.network.Selector.registerChannel(Selector.java:329) ~[na:na]
        at org.apache.kafka.common.network.Selector.connect(Selector.java:256) ~[na:na]
        at org.apache.kafka.clients.NetworkClient.initiateConnect(NetworkClient.java:1032) ~[na:na]
        at org.apache.kafka.clients.NetworkClient.ready(NetworkClient.java:301) ~[na:na]
        at org.apache.kafka.clients.admin.KafkaAdminClient$AdminClientRunnable.sendEligibleCalls(KafkaAdminClient.java:1109) ~[na:na]
        at org.apache.kafka.clients.admin.KafkaAdminClient$AdminClientRunnable.processRequests(KafkaAdminClient.java:1369) ~[na:na]
        at org.apache.kafka.clients.admin.KafkaAdminClient$AdminClientRunnable.run(KafkaAdminClient.java:1312) ~[na:na]
        at java.base@21.0.2/java.lang.Thread.runWith(Thread.java:1596) ~[device-manager:na]
        at java.base@21.0.2/java.lang.Thread.run(Thread.java:1583) ~[device-manager:na]
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:833) ~[device-manager:na]
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:211) ~[na:na]
Caused by: java.security.NoSuchAlgorithmException: class configured for SaslClientFactory (provider: SASL/SCRAM Client Provider) cannot be found.
        at java.base@21.0.2/java.security.Provider$Service.getImplClass(Provider.java:2004) ~[device-manager:na]
        at java.base@21.0.2/java.security.Provider$Service.getDefaultConstructor(Provider.java:2020) ~[device-manager:na]
        at java.base@21.0.2/java.security.Provider$Service.newInstanceOf(Provider.java:1934) ~[device-manager:na]
        at java.base@21.0.2/java.security.Provider$Service.newInstanceUtil(Provider.java:1942) ~[device-manager:na]
        at java.base@21.0.2/java.security.Provider$Service.newInstance(Provider.java:1917) ~[device-manager:na]
        at java.security.sasl@21.0.2/javax.security.sasl.Sasl.loadFactory(Sasl.java:459) ~[na:na]
        ... 23 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.apache.kafka.common.security.scram.internals.ScramSaslClient$ScramSaslClientFactory
        at java.base@21.0.2/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:52) ~[device-manager:na]
        at java.base@21.0.2/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188) ~[na:na]
        at java.base@21.0.2/java.lang.ClassLoader.loadClass(ClassLoader.java:121) ~[device-manager:na]
        at java.base@21.0.2/java.security.Provider$Service.getImplClass(Provider.java:1991) ~[device-manager:na]
        ... 28 common frames omitted
````

**System Info:**
 - OS: Linux
 - GraalVM Version [CE 21.0.2+13.1 (build 21.0.2+13-jvmci-23.1-b30)]
 - Java Version `17`
 - Plugin version: `native-maven-plugin:0.10.1`

--- a/metadata/org.apache.kafka/kafka-clients/3.5.1/reflect-config.json
+++ b/metadata/org.apache.kafka/kafka-clients/3.5.1/reflect-config.json
@@ -141,11 +141,17 @@
   },
   {
     "name": "org.apache.kafka.common.security.scram.internals.ScramSaslClient$ScramSaslClientFactory",
+    "condition": {
+      "typeReachable": "org.apache.kafka.common.security.scram.internals.ScramSaslClientProvider"
+    },
     "allDeclaredConstructors": true,
     "allDeclaredMethods": true
   },
   {
     "name": "org.apache.kafka.common.security.scram.ScramLoginModule",
+    "condition": {
+      "typeReachable": "org.apache.kafka.common.security.authenticator.AbstractLogin"
+    },
     "allDeclaredConstructors": true,
     "allDeclaredMethods": true
   },

--- a/metadata/org.apache.kafka/kafka-clients/3.5.1/reflect-config.json
+++ b/metadata/org.apache.kafka/kafka-clients/3.5.1/reflect-config.json
@@ -140,6 +140,16 @@
     ]
   },
   {
+    "name": "org.apache.kafka.common.security.scram.internals.ScramSaslClient$ScramSaslClientFactory",
+    "allDeclaredConstructors": true,
+    "allDeclaredMethods": true
+  },
+  {
+    "name": "org.apache.kafka.common.security.scram.ScramLoginModule",
+    "allDeclaredConstructors": true,
+    "allDeclaredMethods": true
+  },
+  {
     "name": "org.apache.kafka.common.serialization.BooleanDeserializer",
     "condition": {
       "typeReachable": "org.apache.kafka.clients.consumer.KafkaConsumer"

--- a/tests/src/org.apache.kafka/kafka-clients/3.5.1/build.gradle
+++ b/tests/src/org.apache.kafka/kafka-clients/3.5.1/build.gradle
@@ -18,6 +18,8 @@ dependencies {
     testImplementation 'ch.qos.logback:logback-classic:1.4.5'
     testImplementation 'org.apache.kafka:kafka_2.13:3.5.1'
     testImplementation 'org.apache.kafka:kafka_2.13:3.5.1:test'
+    testImplementation 'org.slf4j:slf4j-api:1.7.36'
+    testImplementation 'org.slf4j:slf4j-simple:1.7.36'
 }
 
 graalvmNative {

--- a/tests/src/org.apache.kafka/kafka-clients/3.5.1/src/test/java/org_apache_kafka/kafka_clients/KafkaClientsTest.java
+++ b/tests/src/org.apache.kafka/kafka-clients/3.5.1/src/test/java/org_apache_kafka/kafka_clients/KafkaClientsTest.java
@@ -286,4 +286,18 @@ class KafkaClientsTest {
         }
     }
 
+    @Test
+    void testSaslScramSHA512() {
+        Map<String, Object> consumerProperties = new HashMap<>();
+        consumerProperties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, KAFKA_SERVER);
+        consumerProperties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, IntegerDeserializer.class);
+        consumerProperties.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        consumerProperties.put(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG, "SASL_SSL");
+        consumerProperties.put(SaslConfigs.SASL_MECHANISM, "SCRAM-SHA-512");
+        consumerProperties.put(SaslConfigs.SASL_JAAS_CONFIG, "org.apache.kafka.common.security.scram.ScramLoginModule required username=admin password=admin-pass;");
+        try (KafkaConsumer consumer = new KafkaConsumer<>(consumerProperties)) {
+            assertThat(consumer).isNotNull();
+        }
+    }
+
 }

